vortex-expr/src/lib.rs

```diff
@@ -36,6 +39,12 @@ pub trait VortexExpr: Debug + Send + Sync + PartialEq<dyn Any> + Display {
     /// Compute result of expression on given batch producing a new batch
     fn evaluate(&self, batch: &ArrayData) -> VortexResult<ArrayData>;
```

Comment:
I wonder if we can some how get a debug_assert where this is used, to compare dtype and evaluation(..).dtype()?

Reply:
I could have implementations implement `_evaluate` and then debug_assert here in `evaluate`, but it doesn't feel _that_ valuable? The mismatch between the returned array and the expression's dtype should be caught pretty quickly, e.g. if they're used to construct a RecordBatch.

Where do you imagine using `dtype`? I'm maybe not being creative enough.

Reply:
When we are transforming expressions we will want to know the type of subexpressions, e.g. for casting. It would be good to know where the expressions type differs from the eval type.

I know that adding debug asserts to compute functions founds loads of bugs, I think the same may be true here.

Reply:
Done.